app/src/main/java/com/github/se/eventradar/viewmodel/qrCode/ScanTicketQrViewModel.kt

```diff
@@ -0,0 +1,132 @@
+  val uiState: StateFlow<QrCodeScanTicketState> = _uiState
+
+  enum class Action {
+    ScanTicket,
+    ApproveEntry,
+    DenyEntry,
+    FirebaseUpdateError,
+    FirebaseFetchError,
+    AnalyserError,
+  }
+
+  enum class Tab {
+    MyEvent,
+    ScanQr
+  }
+
+  private var myEventID: String? =
+      "1" // TODO MUst be chaNged to NUll and initialized by navigation via saveEventID()
+
+  init {
+    qrCodeAnalyser.onTicketDecoded = { decodedString ->
+      val result = decodedString ?: "Failed to decode QR Code"
+      updateDecodedString(result) // Update state flow
+      if (result != "Failed to decode QR Code") {
+        //        println("correctly decoded")
+        updatePermissions(result) // Directly call updateFriendList
+      } else {
+        //      println("wrongly decoded")
+        changeAction(Action.AnalyserError)
+        //        println("checkpoint 2")
+      }
+    }
+  }
+
+  private fun updatePermissions(decodedString: String) {
+    while (myEventID == null) {
+      // Wait until myEventID is not null
+    }
+    println("entered updatePermissions")
+    val uiLength = 28
+    val attendeeID = decodedString.take(uiLength)
+    Log.d("QrCodeFriendViewModel", "Ticket User ID: $attendeeID")
+
+    viewModelScope.launch {
+      val attendeeUserDeferred = async { userRepository.getUser(attendeeID) }
+      val currentEventDeferred = async { eventRepository.getEvent(myEventID!!) }
+
+      val attendeeUser = attendeeUserDeferred.await()
+      val currentEvent = currentEventDeferred.await()
+
+      if (attendeeUser is Resource.Success && currentEvent is Resource.Success) {
+        retryUpdate(attendeeUser.data!!, currentEvent.data!!)
+      } else {
+        changeAction(Action.FirebaseFetchError)
+        return@launch
+      }
+    }
+  }
+
+  private suspend fun retryUpdate(user: User, event: Event) {
+    println("entered retryUpdate")
+    var maxNumberOfRetries = 3
+    var updateResult: Resource<Any>?
+    if (user.eventsAttendeeSet.contains(myEventID) && event.attendeeSet.contains(user.userId)) {
+      println("both contain one another")
+      user.eventsAttendeeSet.remove(myEventID)
+      val userUpdateResult = userRepository.updateUser(user)
+      event.attendeeSet.remove(user.userId)
+      val eventUpdateResult = eventRepository.updateEvent(event)
+
+      do {
+        // Check if both updates were successful
+        updateResult =
+            if (userUpdateResult is Resource.Success && eventUpdateResult is Resource.Success) {
+              changeAction(Action.ApproveEntry)
+              Resource.Success(Unit)
+            } else {
+              changeAction(Action.FirebaseUpdateError)
+              Resource.Failure(Exception("Failed to update user and event"))
+            }
+      } while ((updateResult !is Resource.Success) && (maxNumberOfRetries-- > 0))
+    } else {
+      println("one does not contain the other")
+      changeAction(Action.DenyEntry)
+    }
+  }
```

Comment:
nit: typically a duplicated functionality between the two viewModel. If time, would be nice to have it as a decorator (like coaches suggested) to avoid duplication of code

Reply:
disagree, they do different things.

Reply:
I meant the fact to retry, I know the two functions need to fetch different things